[inductor] Disable fp contraction and add option to use precise division

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122915
* __->__ #115435
* #123269

Fixes #122260

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 committed:
WIP: [inductor] Force IEEE-compliant division in triton

Fixes #101039

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we explicity use the `div_rn` which forces IEEE compliant rounding.

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

bertmaher commented:
Just curious, if we didn't use `fma` here, and instead computed `x*1e6 + y*1e6`, would we still get the wrong result here?  I'm kind of wondering if fma in general is a problem (sort of independently of non-precise division).  I believe the CPU backend for inductor actually disables fma fusion.

peterbell10 commented:
Yeah I'll give that a shot as well.

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

desertfire commented:
I think it will be great if we can provide a torch.compile mode that turns off all the optimizations known to cause large numerical differences. When users find torch.compile produces an unacceptable numerical difference from eager, the first thing they could try is to switch to that special mode.

peterbell10 commented:
> I think it will be great if we can provide a torch.compile mode that turns off all the optimizations known to cause large numerical differences. When users find torch.compile produces an unacceptable numerical difference from eager, the first thing they could try is to switch to that special mode.

I suspect the biggest factor here is that fusions often prevent floating point values from being truncated to half precision. In one sense this is good as it leads to higher precision, but as this issue shows the mismatch in precision between two different paths to compute the same value can lead to numerical issues.

peterbell10 commented:
No significant swings on the benchmark results for disabling fp-contraction. Will have to dig into these torchbench accuracy failures though.

oulgen commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -339,6 +339,7 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
                 "num_warps": compile_meta["num_warps"],
                 "num_stages": compile_meta["num_stages"],
                 "debug": compile_meta["debug"],
+                "enable_fp_fusion": config.triton.enable_fp_fusion,
```
@Chillee pointed out this me. I wonder how this will play with user defined triton kernels? Should we exclude those from this setting?

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Disable fp contraction and add option to use precise division"


Fixes #101039

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Disable fp contraction and add option to use precise division"


Fixes #101039

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -339,6 +339,11 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
+                "enable_fp_fusion": (
+                    config.triton.enable_fp_fusion
+                    or self.heuristic_type
+                    in (HeuristicType.TEMPLATE, HeuristicType.USER_AUTOTUNE)
```
@oulgen @eellison this leaves fma generation enabled for template kernel and user provided kernels.

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Disable fp contraction and add option to use precise division"


Fixes #122260

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Disable fp contraction and add option to use precise division"


Fixes #122260

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
`eca_halonext26ts` may have just been flakey... rebase ?

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Disable fp contraction and add option to use precise division"


Fixes #122260

In the issue we have something like:
```python
(x / 1e-6) + (y / 1e-6)
```
which triton turns into
```python
fma(x, 1e6, y * 1e6)
```

Where we have one division calculated to float32 precision,
and the other calculated to infinite precision as part of the `fma`.
In the issue this leads to a bad normalization in softmax.

Here we disallow the "fp fusion" optimization which generates these fma instructions.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot rebase

eellison reviewed (commented):
checking in - what is latest state of this pr? were you running into any issues and do we still think this is the best approach

peterbell10 committed:
Update

[ghstack-poisoned]

eellison commented:
sounds good!
